plutus-benchmark/script-contexts/src/PlutusBenchmark/V1/ScriptContexts.hs

```diff
@@ -0,0 +1,138 @@
+{-# LANGUAGE BangPatterns      #-}
+{-# LANGUAGE DataKinds         #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TemplateHaskell   #-}
+
+module PlutusBenchmark.V1.ScriptContexts where
+
+import PlutusLedgerApi.V1 (PubKeyHash (..), ScriptContext (..), ScriptPurpose (..), TxId (..),
+                           TxInfo (..), TxOut (..), TxOutRef (..), always)
+import PlutusLedgerApi.V1.Address
+import PlutusLedgerApi.V1.Value
+import PlutusTx qualified
+import PlutusTx.Builtins qualified as PlutusTx
+import PlutusTx.Plugin ()
+import PlutusTx.Prelude qualified as PlutusTx
+
+-- | A very crude deterministic generator for 'ScriptContext's with size
+-- approximately proportional to the input integer.
+mkScriptContext :: Int -> ScriptContext
+mkScriptContext i =
+  ScriptContext
+    (mkTxInfo i)
+    (Spending (TxOutRef (TxId "") 0))
+
+
+mkTxInfo :: Int -> TxInfo
+mkTxInfo i = TxInfo {
+  txInfoInputs=mempty,
+  txInfoOutputs=fmap mkTxOut [1..i],
+  txInfoFee=mempty,
+  txInfoMint=mempty,
+  txInfoDCert=mempty,
+  txInfoWdrl=mempty,
+  txInfoValidRange=always,
+  txInfoSignatories=mempty,
+  txInfoData=mempty,
+  txInfoId=TxId ""
+  }
+
+mkTxOut :: Int -> TxOut
+mkTxOut i = TxOut {
+  txOutAddress=pubKeyHashAddress (PubKeyHash ""),
+  txOutValue=mkValue i,
+  txOutDatumHash=Nothing
+  }
+
+mkValue :: Int -> Value
+mkValue i = assetClassValue (assetClass adaSymbol adaToken) (fromIntegral i)
+
+-- This example decodes the script context (which is O(size-of-context) work), and then also
+-- does some work that's roughly proportional to the size of the script context (counting the
+-- outputs). This should be a somewhat realistic example where a reasonable chunk of work is
+-- done in addition to decoding.
+checkScriptContext1 :: PlutusTx.BuiltinData -> ()
+checkScriptContext1 d =
+  -- Bang pattern to ensure this is forced, probably not necesssary
+  -- since we do use it later
+  let !sc = PlutusTx.unsafeFromBuiltinData d
+      ScriptContext txi _  = sc
+  in
+  if PlutusTx.length (txInfoOutputs txi) `PlutusTx.modInteger` 2 PlutusTx.== 0
```

Comment:
Using field accessors is in general inefficient compared to record patterns, but probably fine if there's only a single usage in a script.